libpam/pam_modutil_sanitize.c: optimize the way to close fds

bessonc opened the pull request:
Using an arbitrary MAX_FD_NO (set to 65535 or 2000000) isn't optimal. This minor patch tries to close only open descriptors beyond STDERR_FILENO thanks to /proc/self/fd. If /proc isn't mounted, fallback to the previous behavior.

t8m commented:
Looks good to me.
@ldv-alt Would you please review it as well?

ldv-alt commented on libpam/pam_modutil_sanitize.c:
```diff
@@ -128,15 +132,30 @@ close_fds(void)
-
-	for (; fd > STDERR_FILENO; --fd)
-		close(fd);
+	/* 
```
trailing whitespace

bessonc replied:
Sorry for that.

ldv-alt commented on libpam/pam_modutil_sanitize.c:
```diff
@@ -128,15 +132,30 @@ close_fds(void)
-		close(fd);
+	/* 
+	 * if /proc is mounted, we can optimize what fd can be closed,
+ 	 * but if it isn't available, fall back to the previous behavior.
```
tab after space

ldv-alt commented on libpam/pam_modutil_sanitize.c:
```diff
@@ -128,15 +132,30 @@ close_fds(void)
+			fd = MIN_FD_NO;
+		else
+			fd = rlim.rlim_max - 1;
+	
```
trailing whitespace

ldv-alt commented on libpam/pam_modutil_sanitize.c:
```diff
@@ -128,15 +132,30 @@ close_fds(void)
+				continue;
+			fd = atoi(dent->d_name);
+			if (fd > STDERR_FILENO)
+				close(fd);
```
real chance to close the file descriptor created by opendir and operated by readdir

bessonc replied:
Yes, the dir descriptor is even always closed in my tests, but it was the last fd. Fixed with a safer code in the last commit. Thx for the patch review.

ldv-alt commented on libpam/pam_modutil_sanitize.c:
```diff
@@ -128,15 +132,30 @@ close_fds(void)
+	 * if /proc is mounted, we can optimize what fd can be closed,
+ 	 * but if it isn't available, fall back to the previous behavior.
+	 */
+	if ((dir = opendir("/proc/self/fd")) != NULL) {
```
what if /proc/self/fd is not a valid proc filesystem?

bessonc replied:
I'm confused. Several parts of pam code relies on /proc, and I can't see any check if /proc is a valid proc fs. I'm not against a change on that, but please explain me how to check if /proc is a valid fs? Using functions like getmntent (which can check a given mnt_dir and mnt_type) reads /etc/mtab, which is itself a symlink to /proc/self/mounts...

t8m replied:
I would not really care about this. So severally bogus setup is not something we want to support anyway. /proc must be procfs or anything can happen.

ldv-alt replied:
I don't want any regressions in case of /proc is not mounted, so I suggest adding a statfs(dirfd(dir)) and checking that f_type matches PROC_SUPER_MAGIC.

Ideally, we could use close_range syscall when linux finally implements it, but we are not there yet.

ldv-alt replied:
The check for PROC_SUPER_MAGIC would require AC_CHECK_HEADERS(linux/magic.h),
#include <linux/magic.h> under #ifdef HAVE_LINUX_MAGIC_H,
and the whole check would have to be guarded with #ifdef PROC_SUPER_MAGIC.

t8m replied:
But why there would be any regression? If /proc is not mounted there won't be /proc/self/fd and so the old code will be used and if somebody with sufficient rights makes an effort to break his environment to create some bogus thing in /proc/self/fd without mounting /proc I do not know why he should not bear its consequences. He can as well do bind mount from some another procfs directory to /proc/self/fd and you would not be able to detect it even with this trick.

bessonc replied:
I've just pushed a new commit implementing this check. Would you please review it and let me know if I have to squash my two commits into a single one.

ldv-alt commented on libpam/pam_modutil_sanitize.c:
```diff
@@ -128,15 +132,30 @@ close_fds(void)
+			if (!strcmp(dent->d_name, ".")
+			    || !strcmp(dent->d_name, ".."))
+				continue;
+			fd = atoi(dent->d_name);
```
note that atoi(".") == atoi("..") == 0

ldv-alt commented on libpam/pam_modutil_sanitize.c:
```diff
@@ -128,15 +132,30 @@ close_fds(void)
+	
+		for (; fd > STDERR_FILENO; --fd)
+			close(fd);
+	}
```
The dir structure created by opendir is definitely leaked at this point.
If the file descriptor created by opendir is less or equal to STDERR_FILENO, it's also leaked.

bessonc replied:
Should not happen after the refactoring. But I think that the fd for opendir() can't be less or equal to STDERR_FILENO.

bessonc commented:
The last commit should be safer. It results to the following syscalls:
`10196 04:09:50.068538 openat(AT_FDCWD, "/proc/self/fd", O_RDONLY|O_NONBLOCK|O_CLOEXEC|O_DIRECTORY) = 6</proc/10196/fd> <0.000012>
10196 04:09:50.068594 fstat(6</proc/10196/fd>, {st_dev=makedev(0, 4), st_ino=65069, st_mode=S_IFDIR|0500, st_nlink=2, st_uid=1004, st_gid=1004, st_blksize=1024, st_blocks=0, st_size=0, st_atime=1565338190 /* 201
9-08-09T04:09:50.066920028-0400 */, st_atime_nsec=66920028, st_mtime=1565338190 /* 2019-08-09T04:09:50.066920028-0400 */, st_mtime_nsec=66920028, st_ctime=1565338190 /* 2019-08-09T04:09:50.066920028-0400 */, st_
ctime_nsec=66920028}) = 0 <0.000006>
10196 04:09:50.068664 getdents64(6</proc/10196/fd>, [{d_ino=65069, d_off=1, d_reclen=24, d_type=DT_DIR, d_name="."}, {d_ino=65068, d_off=2, d_reclen=24, d_type=DT_DIR, d_name=".."}, {d_ino=65071, d_off=3, d_recl
en=24, d_type=DT_LNK, d_name="0"}, {d_ino=65075, d_off=4, d_reclen=24, d_type=DT_LNK, d_name="1"}, {d_ino=65076, d_off=5, d_reclen=24, d_type=DT_LNK, d_name="2"}, {d_ino=65077, d_off=6, d_reclen=24, d_type=DT_LN
K, d_name="3"}, {d_ino=65070, d_off=7, d_reclen=24, d_type=DT_LNK, d_name="4"}, {d_ino=65078, d_off=8, d_reclen=24, d_type=DT_LNK, d_name="5"}, {d_ino=65073, d_off=66, d_reclen=24, d_type=DT_LNK, d_name="6"}], 3
2768) = 216 <0.000018>
10196 04:09:50.068724 close(3</var/lib/sss/mc/passwd>) = 0 <0.000007>
10196 04:09:50.068755 close(4<pipe:[65066]>) = 0 <0.000006>
10196 04:09:50.068782 close(5<pipe:[65066]>) = 0 <0.000006>
10196 04:09:50.068809 getdents64(6</proc/10196/fd>, [], 32768) = 0 <0.000006>
10196 04:09:50.068839 close(6</proc/10196/fd>) = 0 <0.000006>`

Instead of:
`23934 11:22:37.718450 prlimit64(0, RLIMIT_NOFILE, NULL, {rlim_cur=5120*1024, rlim_max=5120*1024}) = 0 <0.000013>
23934 11:22:37.718493 close(65535)      = -1 EBADF (Bad file descriptor) <0.000013>
23934 11:22:37.718539 close(65534)      = -1 EBADF (Bad file descriptor) <0.000013>
23934 11:22:37.718582 close(65533)      = -1 EBADF (Bad file descriptor) <0.000013>
23934 11:22:37.718624 close(65532)      = -1 EBADF (Bad file descriptor) <0.000013>
...
...
23934 11:22:39.675547 close(9)          = -1 EBADF (Bad file descriptor) <0.000010>
23934 11:22:39.675577 close(8)          = -1 EBADF (Bad file descriptor) <0.000010>
23934 11:22:39.675607 close(7)          = -1 EBADF (Bad file descriptor) <0.000010>
23934 11:22:39.675640 close(6)          = -1 EBADF (Bad file descriptor) <0.000010>
23934 11:22:39.675671 close(5<pipe:[162650]>) = 0 <0.000013>
23934 11:22:39.675710 close(4<pipe:[162650]>) = 0 <0.000008>
23934 11:22:39.675737 close(3</var/lib/sss/mc/passwd>) = 0 <0.000009>`

t8m commented:
@bessonc Could you please squash the two commits and force-push?

ldv-alt commented on libpam/pam_modutil_sanitize.c:
```diff
@@ -116,27 +117,46 @@ redirect_out(pam_handle_t *pamh, enum pam_modutil_redirect_fd mode,
+			}
+		}
+	}
+	closedir(dir);
```
Let's not  guess whether closedir(NULL) is safe and move the closedir invocation into dir != NULL branch instead.

ldv-alt commented on libpam/pam_modutil_sanitize.c:
```diff
@@ -116,27 +117,46 @@ redirect_out(pam_handle_t *pamh, enum pam_modutil_redirect_fd mode,
+	closedir(dir);
+
+	/* If /proc isn't available, fallback to the previous behavior. */
+	if (!fd) {
```
Let's not depend on the order of directory entries and check for dfd >= 0 instead.
Of course it means that dfd would have to be initialized to -1.

ldv-alt commented:
On Tue, Aug 20, 2019 at 07:22:24AM -0700, bessonc wrote:
> bessonc commented on this pull request.
> 
> > @@ -128,15 +132,30 @@ close_fds(void)
>  	int fd;
>  	struct rlimit rlim;
>  
> -	if (getrlimit(RLIMIT_NOFILE, &rlim) || rlim.rlim_max > MAX_FD_NO)
> -		fd = MAX_FD_NO;
> -	else if (rlim.rlim_max < MIN_FD_NO)
> -		fd = MIN_FD_NO;
> -	else
> -		fd = rlim.rlim_max - 1;
> -
> -	for (; fd > STDERR_FILENO; --fd)
> -		close(fd);
> +	/* 
> +	 * if /proc is mounted, we can optimize what fd can be closed,
> + 	 * but if it isn't available, fall back to the previous behavior.
> +	 */
> +	if ((dir = opendir("/proc/self/fd")) != NULL) {
> 
> I've just pushed a new commit implementing this check. Would you please review it and let me know if I have to squash my two commits into a single one.

You've already got a descriptor, so lease use fstatfs instead of statfs,
and yes, please squash your two commits into a single one.

Thanks

bessonc commented:
> You've already got a descriptor, so lease use fstatfs instead of statfs, and yes, please squash your two commits into a single one.

I had a doubt... Fixed and rebased.
Thanks!

ldv-alt commented:
On Thu, Aug 22, 2019 at 03:13:49AM -0700, bessonc wrote:
> > You've already got a descriptor, so lease use fstatfs instead of statfs, and yes, please squash your two commits into a single one.
> 
> I had a doubt... Fixed and rebased.
> Thanks!

There are two issues.

First, sys/vfs.h is added to AC_CHECK_HEADERS but it's included unconditionally.

Second, dfd is not reset to -1 when it doesn't pass the is_in_procfs check,
defeating the (dfd < 0) check.

Thanks

bessonc commented:
> There are two issues. First, sys/vfs.h is added to AC_CHECK_HEADERS but it's included unconditionally.

Added an ifdef HAVE_SYS_VFS_H to include sys/vfs.h.

> Second, dfd is not reset to -1 when it doesn't pass the is_in_procfs check, defeating the (dfd < 0) check. Thanks

Fixed.

Thanks

ldv-alt commented:
On Fri, Aug 23, 2019 at 06:30:46AM -0700, bessonc wrote:
> > There are two issues. First, sys/vfs.h is added to AC_CHECK_HEADERS but it's included unconditionally.
> 
> Added an ifdef HAVE_SYS_VFS_H to include sys/vfs.h.

I think "#ifdef PROC_SUPER_MAGIC" has to be changed to "#if defined HAVE_SYS_VFS_H && defined PROC_SUPER_MAGIC".

Besides that, the change looks fine now, thanks.


-- 
ldv

bessonc commented:
> I think "#ifdef PROC_SUPER_MAGIC" has to be changed to "#if defined HAVE_SYS_VFS_H && defined PROC_SUPER_MAGIC". Besides that, the change looks fine now, thanks.

I followed your recommendations.

ldv-alt commented:
On Mon, Aug 26, 2019 at 01:10:42AM -0700, bessonc wrote:
> > I think "#ifdef PROC_SUPER_MAGIC" has to be changed to "#if defined HAVE_SYS_VFS_H && defined PROC_SUPER_MAGIC". Besides that, the change looks fine now, thanks.
> 
> I followed your recommendations.

I think this edition is fine.  Tomáš?

t8m commented:
I am OK with it so I am going to merge it.
